assertions/scanner.go

```diff
@@ -419,35 +423,67 @@ func (m *Manager) findLastAgreedWithAncestor(
+func (m *Manager) keepTryingAssertionConfirmation(ctx context.Context, assertionHash protocol.AssertionHash) {
 	for {
+		if m.assertionConfirmed(ctx, assertionHash) {
+			return
+		}
+		ticker := time.NewTicker(m.confirmationAttemptInterval)
 		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
 		case <-ticker.C:
-			status, err := s.chain.AssertionStatus(ctx, assertionHash)
+		}
+	}
+}
+
+func (m *Manager) assertionConfirmed(ctx context.Context, assertionHash protocol.AssertionHash) bool {
+	status, err := m.chain.AssertionStatus(ctx, assertionHash)
+	if err != nil {
+		srvlog.Error("Could not get assertion by hash", log.Ctx{"err": err, "assertionHash": assertionHash.Hash})
+		return false
+	}
+	if status == protocol.NoAssertion {
+		srvlog.Error("No assertion found by hash", log.Ctx{"err": err, "assertionHash": assertionHash.Hash})
+		return false
+	}
+	if status == protocol.AssertionConfirmed {
+		srvlog.Info("Assertion confirmed", log.Ctx{"assertionHash": assertionHash.Hash})
+		return true
+	}
+	err = m.chain.ConfirmAssertionByTime(ctx, assertionHash)
+	if err != nil {
+		if strings.Contains(err.Error(), protocol.BeforeDeadlineAssertionConfirmationError) {
+			creationInfo, err := m.chain.ReadAssertionCreationInfo(ctx, assertionHash)
 			if err != nil {
-				srvlog.Error("Could not get assertion by hash", log.Ctx{"err": err, "assertionHash": assertionHash.Hash})
-				continue
-			}
-			if status == protocol.NoAssertion {
-				srvlog.Error("No assertion found by hash", log.Ctx{"err": err, "assertionHash": assertionHash.Hash})
-				continue
+				srvlog.Error("Could not get assertion creation info by hash", log.Ctx{"err": err, "assertionHash": assertionHash.Hash})
+				return false
 			}
-			if status == protocol.AssertionConfirmed {
-				srvlog.Info("Assertion confirmed", log.Ctx{"assertionHash": assertionHash.Hash})
-				return
+			prevCreationInfo, err := m.chain.ReadAssertionCreationInfo(ctx, protocol.AssertionHash{Hash: creationInfo.ParentAssertionHash})
+			if err != nil {
+				srvlog.Error("Could not get assertion creation info by hash", log.Ctx{"err": err, "assertionHash": creationInfo.ParentAssertionHash})
+				return false
 			}
-			err = s.chain.ConfirmAssertionByTime(ctx, assertionHash)
+			latestHeader, err := m.chain.Backend().HeaderByNumber(ctx, nil)
 			if err != nil {
-				continue
+				srvlog.Error("Could not get latest header", log.Ctx{"err": err})
+				return false
 			}
-			srvlog.Info("Assertion confirmed", log.Ctx{"assertionHash": assertionHash.Hash})
-			return
-		case <-ctx.Done():
-			return
+			<-time.After(m.getConfirmationInterval(creationInfo.CreationBlock, prevCreationInfo.ConfirmPeriodBlocks, latestHeader.Number.Uint64()))
 		}
+		return false
+	}
+	srvlog.Info("Assertion confirmed", log.Ctx{"assertionHash": assertionHash.Hash})
+	return true
+}
+
+func (m *Manager) getConfirmationInterval(creationBlock uint64, confirmPeriodBlocks uint64, latestHeaderNumber uint64) time.Duration {
```

Comment:
Thanks @amsanghi for this PR! I'll merge as is, but would be good in a follow-up to add more unit tests to this particular helper. It's very important and want to make sure the math is what we expect